test/client/connection.spec.js

```diff
@@ -0,0 +1,317 @@
+        form: '<form class="or"><img src="/path/to/source.png"/></form>',
+        model: '<model/>',
+        hash: '12345',
+    } );
+
+    describe( 'Uploading records', () => {
+
+        /** @type { SinonSandbox } */
+        let sandbox;
+
+        /** @type {Survey} */
+        let survey;
+
+        /** @type { EnketoRecord } */
+        let record;
+
+        /** @type { StubbedRequest[] } */
+        let requests;
+
+        /** @type { window.Response } */
+        let response;
+
+        const stubSuccessRespopnse = () => {
+            response = {
+                ok: true,
+                status: 201,
+                text() {
+                    return Promise.resolve( `
+                        <OpenRosaResponse xmlns="http://openrosa.org/http/response">
+                            <message nature="submit_success">Success</message>
+                        </OpenRosaResponse>
+                    ` );
+                },
+            };
+        };
+
+        beforeEach( () => {
+            response = {
+                status: 500,
+                text() {
+                    return Promise.resolve( '<error>No stub response designated by test</error>' );
+                },
+            };
+
+            requests = [];
+
+            survey = Object.assign( {}, baseSurvey );
+
+            record = {
+                enketoId,
+                instanceId,
+                name: 'name A',
+                xml: '<model><something>a</something></model>',
+                files: [],
+            };
+
+            sandbox = sinon.createSandbox();
+            sandbox.stub( settings, 'enketoId' ).get( () => enketoId );
+
+            sandbox.stub( window, 'fetch' ).callsFake( ( url, init ) => {
+                requests.push( { url, init } );
+
+                return Promise.resolve( response );
+            } );
+        } );
+
+        afterEach( () => {
+            sandbox.restore();
+        } );
+
+        it( 'uploads a record', done => {
+            stubSuccessRespopnse();
+
+            connection.uploadRecord( record )
+                .then( result => {
+                    expect( result.status ).to.equal( 201 );
+                    expect( requests.length ).to.equal( 1 );
+
+                    const request = requests[0];
+                    const body = Object.fromEntries( request.init.body.entries() );
+                    const instanceId = request.init.headers['X-OpenRosa-Instance-Id'];
+                    const submission = body.xml_submission_file;
+
+                    expect( instanceId ).to.equal( record.instanceId );
+                    expect( submission instanceof File ).to.equal( true );
+
+                    return submission.text();
+                } )
+                .then( submission => {
+                    expect( submission ).to.equal( record.xml );
+                } )
+                .then( done, done );
+        } );
+
+        describe( 'last-saved records', () => {
```

Comment:
Some of these feel like they'd also be good candidates for pulling out to a common feature test place. But this might really be a matter of taste.